actions/aqua-security-dependency/main.go

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2018-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/paketo-buildpacks/pipeline-builder/actions"
+)
+
+func main() {
+	inputs := actions.NewInputs()
+
+	u, ok := inputs["username"]
+	if !ok {
+		panic(fmt.Errorf("username must be specified"))
+	}
+
+	p, ok := inputs["password"]
+	if !ok {
+		panic(fmt.Errorf("password must be specified"))
+	}
+
+	uri := "https://get.aquasec.com/scanner-releases.txt"
+
+	resp, err := http.Get(uri)
+	if err != nil {
+		panic(fmt.Errorf("unable to get %s\n%w", uri, err))
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		panic(fmt.Errorf("unable to download %s: %d", uri, resp.StatusCode))
+	}
+
+	versions := make(actions.Versions)
+
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
+		v := scanner.Text()
+		versions[v] = fmt.Sprintf("https://download.aquasec.com/scanner/%s/scannercli", v)
+	}
+	if err := scanner.Err(); err != nil {
+		panic(fmt.Errorf("unable to read index\n%w", err))
+	}
+
+	versions.GetLatestWithRequest(inputs, func(req *http.Request) *http.Request {
```

Comment:
Also, I am confused about the pattern here regarding `inputs`. Is each `main` function or is `GetLatestWithRequest` responsible for reading and interpreting inputs?

What inputs are common? Is it just `pre_release`? If `GetLatestWithRequest ` is responsible for understanding the allowed values of inputs, should it look for `username` and `password` inputs and modify the request on its own instead of accepting a function? If the `main` func is responsible, should we just be passing in a `preRelease` param, instead of passing in all of the inputs?

Reply:
Both are true.  There's a single `inputs` object (mapping to the `with:` directive in an action) that both the individual actions extract their configuration from, as the `Latest` selector pulling common configuration from.  The collection of "common" information has changed over time (specifically through the Concourse resource era), enough that keeping it abstract by passing all config through and letting the `Latest` selector pull what it needs, has proven to be more manageable.